File: kreartif-api/controllers/komentarController.js
const { Komentar } = require('../models/komentarModel');
const { Karya } = require('../models/karyaModel');
const { Notifikasi } = require('../models/notifikasiModel');
const { User } = require('../models/userModel');

// GET: Semua komentar untuk suatu karya
exports.getKomentarByKarya = async (req, res) => {
  try {
    const { karya_id } = req.params;
    const komentar = await Komentar.findAll({
      where: { karya_id },
      include: [
        {
          model: User,
          attributes: ['username', 'profile_picture'], 
        },
      ],
      order: [['createdAt', 'DESC']],
    });
    res.json(komentar);
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
};

// POST: Tambah komentar + notifikasi ke pemilik karya
exports.addKomentar = async (req, res) => {
  try {
    const { karya_id } = req.params;
    const { isi } = req.body;

    const karya = await Karya.findByPk(karya_id);
    if (!karya) return res.status(404).json({ msg: 'Karya tidak ditemukan' });

    const komentar = await Komentar.create({
      karya_id,
      user_id: req.user.id,
      isi
    });

    // Tambah notifikasi jika bukan mengomentari karya sendiri
    if (req.user.id !== karya.user_id) {
      await Notifikasi.create({
        user_id: karya.user_id,       
        dari_user_id: req.user.id,    
        karya_id,
        jenis: 'komentar',
      });
    }

    res.status(201).json(komentar);
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
};

// GET: Hitung komentar untuk karya tertentu
exports.getKomentarCount = async (req, res) => {
  try {
    const { karya_id } = req.params;
    const count = await Komentar.count({ where: { karya_id } });
    res.json({ count });
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
};

File: kreartif-api/controllers/komentarController.py
from fastapi import HTTPException
from sqlalchemy.orm import Session, joinedload
from models.komentarModel import Komentar
from models.karyaModel import Karya
from models.notifikasiModel import Notifikasi
from models.userModel import User

# GET: Semua komentar untuk suatu karya
def get_komentar_by_karya(karya_id: int, db: Session):
    try:
        komentar_list = (
            db.query(Komentar)
            .filter(Komentar.karya_id == karya_id)
            .options(joinedload(Komentar.user).load_only(User.username, User.profile_picture))
            .order_by(Komentar.created_at.desc())
            .all()
        )
        return komentar_list
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

# POST: Tambah komentar + notifikasi ke pemilik karya
def add_komentar(karya_id: int, komentar_data, db: Session, current_user: dict):
    try:
        isi = komentar_data.isi
        karya = db.query(Karya).filter(Karya.id == karya_id).first()

        if not karya:
            raise HTTPException(status_code=404, detail="Karya tidak ditemukan")

        komentar = Komentar(
            karya_id=karya_id,
            user_id=current_user["id"],
            isi=isi
        )
        db.add(komentar)
        db.commit()
        db.refresh(komentar)

        if current_user["id"] != karya.user_id:
            notifikasi = Notifikasi(
                user_id=karya.user_id,
                dari_user_id=current_user["id"],
                karya_id=karya_id,
                jenis="komentar"
            )
            db.add(notifikasi)
            db.commit()

        return komentar
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

# GET: Hitung komentar untuk karya tertentu
def get_komentar_count(karya_id: int, db: Session):
    try:
        count = db.query(Komentar).filter(Komentar.karya_id == karya_id).count()
        return {"count": count}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
